relay-wstring/src/lib.rs

```diff
@@ -0,0 +1,509 @@
+    /// the subslice is not on character boundaries or otherwise invalid this will return
+    /// `None`.
+    pub fn get<I>(&self, index: I) -> Option<&<I as SliceIndex<WStr>>::Output>
+    where
+        I: SliceIndex<WStr>,
+    {
+        index.get(self)
+    }
+
+    /// Return a mutable subslice of `Self`.
+    ///
+    /// The slice indices are on byte offsets of the underlying UTF-16LE encoded buffer, if
+    /// the subslice is not on character boundaries or otherwise invalid this will return
+    /// `None`.
+    pub fn get_mut<I>(&mut self, index: I) -> Option<&mut <I as SliceIndex<WStr>>::Output>
+    where
+        I: SliceIndex<WStr>,
+    {
+        index.get_mut(self)
+    }
+
+    /// Return a subslice of `Self`.
+    ///
+    /// # Safety
+    ///
+    /// Like [Self::get] but this results in undefined behaviour if the sublice is not on
+    /// character boundaries or otherwise invalid.
+    pub unsafe fn get_unchecked<I>(&self, index: I) -> &<I as SliceIndex<WStr>>::Output
+    where
+        I: SliceIndex<WStr>,
+    {
+        index.get_unchecked(self)
+    }
+
+    /// Return a mutable subslice of `Self`.
+    ///
+    /// # Safety
+    ///
+    /// Lice [Self::get_mut] but this results in undefined behaviour if the subslice is not
+    /// on character boundaries or otherwise invalid.
+    pub unsafe fn get_unchecked_mut<I>(&mut self, index: I) -> &mut <I as SliceIndex<WStr>>::Output
+    where
+        I: SliceIndex<WStr>,
+    {
+        index.get_unchecked_mut(self)
+    }
+
+    /// Returns an iterator of the [char]s of a string slice.
+    pub fn chars(&self) -> WStrChars {
+        WStrChars {
+            chunks: self.raw.chunks_exact(2),
+        }
+    }
+
+    /// Returns and iterator over the [char]s of a string slice and their positions.
+    pub fn char_indices(&self) -> WStrCharIndices {
+        WStrCharIndices {
+            chars: self.chars(),
+            index: 0,
+        }
+    }
+
+    /// Returns the [WStr] as a new owned [String].
+    pub fn to_utf8(&self) -> String {
+        self.chars().collect()
+    }
+
+    /// Returns `true` if all characters in the string are ASCII.
+    pub fn is_ascii(&self) -> bool {
+        self.as_bytes().is_ascii()
+    }
+}
+
+/// Iterator yielding `char` from a UTF-16 little-endian encoded byte slice.
+///
+/// The slice must contain valid UTF-16, otherwise this may panic or cause undefined
+/// behaviour.
+#[derive(Debug)]
+pub struct WStrChars<'a> {
+    chunks: ChunksExact<'a, u8>,
+}
+
+impl<'a> Iterator for WStrChars<'a> {
+    type Item = char;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Our input is valid UTF-16LE, so we can take a lot of shortcuts.
+        let chunk = self.chunks.next()?;
+        let mut buf: [u8; 2] = [0; 2]; // TODO: avoid init
+        buf.copy_from_slice(chunk);
+        let u = u16::from_le_bytes(buf);
+
+        if u < 0xD800 || 0xDFFF < u {
+            Some(unsafe { std::char::from_u32_unchecked(u as u32) })
```

Comment:
Can't we use `from_u32` and catch the error? `from_u32` seems to almost have the same condition as you write here

```rust
std::char::from_u32(u as u32).unwrap_or_else(|| {
    // prev else branch here
})
```

Reply:
hmm, i thought this was a neat idea, but after trying it i actually dislike it.  i think it's a case of "explicit is better than implicit" and since we do the decoder math in the else branch already to me it's clearer with the current if branch

Reply:
can we do `from_u32().unwrap()` then? It should be optimized away

Reply:
@untitaker argued that he prefers panics (unwrap) over unsafe so i followed this suggestion anyway.  there's one more if branch and an unwrap now.  i'm guessing this doesn't really matter performance wise

Reply:
Hm, I disagree with this. This is very central code in hot loops. If we can statically infer that this doesn't fail, then let's use unsafe over panics, or find a way to avoid the unsafety to begin with.

Reply:
avoiding the unsafety was my first suggestion

Reply:
So shall I revert this back to the unsafe version, perhaps with some more comments?

Reply:
I think it's fine now